src/keytree.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) 2015 Pieter Wuille
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <vector>
+
+#include "hash.h"
+#include "keytree.h"
+#include "pubkey.h"
+#include "crypto/sha256.h"
+
+#include <secp256k1.h>
+
+extern secp256k1_context_t* secp256k1_bitcoin_verify_context;
+
+namespace {
+
+/* This implements a constant-space merkle root/branch calculator, limited to 2^63 leaves. */
+template<typename LeafSource>
+void MerkleComputation(LeafSource& source, uint256* proot, std::vector<uint256>* pbranch) {
+    if (pbranch) pbranch->clear();
+    if (!source) {
+        if (proot) *proot = uint256();
+        return;
```

Comment:
I'd much prefer a method with a name like exhausted(), rather than a cast to bool.